### DISCUSSION TITLE
Validating and then transforming data with templating

### DISCUSSION THREAD

**Discussion by jmgilman** [cue-lang/cue#1906] [Q&A]
Context: https://gist.github.com/jmgilman/39acdb7d2a9167f697777e845d65f592

This is a very simplified example but should be sufficient to get my question across. I have some JSON data that I need to validate and then transform using CUE's templating feature. Refer to `test.cue` and `test.json` respectively in the gist. When I attempt to export the transformed data, I get:

```
❯ cue export --expression rendered --out text test.cue test.json 
error in call to text/template.Execute: cannot convert non-concrete value string:
    ./test.cue:15:11
    ./test.cue:5:2
```

I think I can understand that `#Config` is not concrete, but how else would I reference the incoming data structure?

**Comment by verdverm** [comment]
The issue is before template rendering, in understanding where `cue` puts the json data.
- `#Config` is the schema, which is not where the JSON ends up
- The JSON ends up at the top level if you don't configure it otherwise
- add `#Config` as an embed to enforce the schema on JSON without placing it somewhere
- you cannot place JSON at a location in the CUE today, though you can place it based on a key (--path) `cue export test.json -l key1`

But this won't work for your case, since you cannot pass the root value to the template function. What you can do is modify your data to have a top-level `config` field and add `config: #Config` to your test.cue and then pass config to template.

You might also be interested in my `hof` project https://github.com/hofstadter-io/hof and the `hof gen` command which combines CUE, data, and text/template into a purpose built code generation tool. For your example, you should be able to run `hof gen test.cue test.json -T test.txt` if you put the template content in `test.txt` and remove some lines from the `test.cue` file / make one of the data+schema alignments from the list above.
